### PR TITLE
[PB-620 PB-659]: feat/stop-remove sharing

### PR DESCRIPTION
**Pull request by miguelsw**
- About AFS
- Add two new EPs to stop and removing from shared folder, will be used on stop share button and remove permission of one user

### CODE REVIEW

File: src/modules/private-share-folder/private-sharing.usecase.ts [hunk 1]
```diff
@@ -2,22 +2,30 @@ import { Injectable } from '@nestjs/common';
...
   }
 }
+export class FolderNotSharedError extends Error {
+  constructor() {
+    super('This folder is not shared');
+  }
+}
+export class UserNotInSharedFolder extends Error {
+  constructor() {
+    super('User is not in shared folder');
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.usecase.ts:21]: same here

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  ) {}
+
+  private async removeByField(fields: Partial<PrivateSharingFolderRole>) {
+    const privateFolder = await this.privateSharingFolderRolesModel.destroy({
+      where: fields,
+    });
+    return privateFolder;
+  }
+
+  async removeByUser(folderUuid: string, userUuid: string): Promise<any> {
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:37]: You must match the typing here with what is specified in the interface.

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    return await this.removeByField({ folderId: folderUuid, userId: userUuid });
+  }
+
+  async removeByFolder(folderUuid: string): Promise<any> {
+    return await this.removeByField({ folderId: folderUuid });
+  }
+
+  async findByFolder(folderUuid: string): Promise<PrivateSharingFolderRole[]> {
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:45]: You must match the typing here with what is specified in the interface.

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      where: fields,
+    });
+    return privateFolder;
+  }
+
+  async removeByUser(folderUuid: string, userUuid: string): Promise<any> {
+    return await this.removeByField({ folderId: folderUuid, userId: userUuid });
+  }
+
+  async removeByFolder(folderUuid: string): Promise<any> {
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:41]: You must match the typing here with what is specified in the interface.

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  async findByFolder(folderUuid: string): Promise<PrivateSharingFolderRole[]> {
+    const sharedFolderRolesByFolder =
+      await this.privateSharingFolderRolesModel.findAll({
+        where: { folderId: folderUuid },
+      });
+    return sharedFolderRolesByFolder;
+  }
+
+  async findByFolderAndUser(
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:53]: You must match the typing here with what is specified in the interface.

File: src/modules/private-share-folder/private-sharing.controller.ts [hunk 6]
```diff
@@ -239,4 +242,55 @@ export class PrivateSharingController {
...
+  @ApiBearerAuth()
+  async stopSharing(
+    @Param('folderId', ParseUUIDPipe) folderUuid: string,
+    @UserDecorator() user: User,
+  ): Promise<any> {
+    try {
+      return await this.privateSharingUseCase.stopSharing(folderUuid);
+    } catch (error) {
+      new Logger().error(
+        `[PRIVATESHARING/STOP] ERROR: ${
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.controller.ts:259]: You must catch the specific errors defined in the use case and handle them. This error for example:

https://github.com/internxt/drive-server-wip/blob/59fe6c7bbd4c8732691028b6491a26f714655c21/src/modules/private-share-folder/private-sharing.usecase.ts#L101

how to do it :
https://github.com/internxt/drive-server-wip/blob/59fe6c7bbd4c8732691028b6491a26f714655c21/src/modules/folder/folder.controller.ts#L304

File: src/modules/private-share-folder/private-sharing.controller.ts [hunk 7]
```diff
@@ -239,4 +242,55 @@ export class PrivateSharingController {
...
+    @UserDecorator() user: User,
+  ): Promise<any> {
+    try {
+      return await this.privateSharingUseCase.removeUserShared(
+        folderUuid,
+        userUuid,
+      );
+    } catch (error) {
+      new Logger().error(
+        `[PRIVATESHARING/REMOVE] ERROR: ${
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.controller.ts:286]: same here

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 8]
```diff
@@ -31,12 +41,32 @@ export class SequelizePrivateSharingRepository
...
   ) {}
 
+  private async removeByField(fields: Partial<PrivateSharingFolder>) {
+    const privateFolder = await this.privateSharingFolderModel.destroy({
+      where: fields,
+    });
+    return privateFolder;
+  }
+
+  async removeByFolderUuid(folderUuid: string): Promise<any> {
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.repository.ts:55]: check type

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 9]
```diff
@@ -31,12 +41,32 @@ export class SequelizePrivateSharingRepository
...
+    const privatesRemovedByFolderUuid = await this.removeByField({
+      folderId: folderUuid,
+    });
+    return privatesRemovedByFolderUuid;
+  }
+
+  async removeBySharedWith(folderUuid: string, userUuid: string): Promise<any> {
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.repository.ts:62]: same here

File: src/modules/private-share-folder/private-sharing.usecase.ts [hunk 10]
```diff
@@ -2,22 +2,30 @@ import { Injectable } from '@nestjs/common';
...
+import { PrivateSharingFolderRolesRepository } from './private-sharing-folder-roles.repository';
 
 export class InvalidOwnerError extends Error {
   constructor() {
     super('You are not the owner of this folder');
   }
 }
+export class FolderNotSharedError extends Error {
+  constructor() {
+    super('This folder is not shared');
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.usecase.ts:16]: You must add 

Object.setPrototypeOf(this, FolderNotSharedError .prototype)

this ensures that the created error instance has the correct prototype, and the proper functioning of instanceof in the controller's catch

File: src/modules/private-share-folder/private-sharing.usecase.ts [hunk 11]
```diff
@@ -70,4 +78,68 @@ export class PrivateSharingUseCase {
...
+      await this.privateSharingRespository.removeBySharedWith(
+        folderUuid,
+        userUuid,
+      );
+    const removed = folderRolesRemoved + userSharedRemoved > 0;
+    return { removed };
+  }
+
+  private async ValidateUserInFolderShared(
+    folderUuid: string,
```
Comment (cg27shokworks) [inline src/modules/private-share-folder/private-sharing.usecase.ts:125]: check types

**Review by cg27shokworks — CHANGES_REQUESTED** [review]

File: src/modules/folder/folder.usecase.ts [hunk 12]
```diff
@@ -24,6 +24,12 @@ const invalidName = /[\\/]|^\s*$/;
 
 type SortParams = Array<[SortableFolderAttributes, 'ASC' | 'DESC']>;
 
+export class NotFoundFolderException extends NotFoundException {
+  constructor() {
```
Comment (sg-gs) [inline src/modules/folder/folder.usecase.ts:28]: Please use [this](https://github.com/internxt/drive-server-wip/blob/cc313904b564f55293fb6d0ce1b2482acf548ead/src/modules/user/user.usecase.ts#L47) so the `instanceof` check works properly

File: src/modules/folder/folder.usecase.ts [hunk 13]
```diff
@@ -55,6 +61,16 @@ export class FolderUseCases {
     return folder;
   }
 
+  async getFolderByUuid(folderUuid: FolderAttributes['uuid']): Promise<Folder> {
```
Comment (sg-gs) [inline src/modules/folder/folder.usecase.ts:64]: This method already [exists](https://github.com/internxt/drive-server-wip/blob/master/src/modules/folder/folder.usecase.ts#L41). 

Also, I do not see this function being used on any of the introduced changes. If it is going to be used, use the already existing one.

If this is needed for accessing a folder by an invited user, you should: 
- Create a use case on private sharing folder use-cases (`getFolderByInvitedUser` p.e) where you check that the user has the proper permissions to access it, and then you pass to the already existent function here, the owner user object.

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      where: fields,
+    });
+    return amountRemoves;
+  }
+
+  async removeByUser(
+    folderUuid: Folder['uuid'],
+    userUuid: User['uuid'],
+  ): Promise<number> {
+    return await this.removeByField({ folderId: folderUuid, userId: userUuid });
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:43]: return directly and remove `async-await` from here

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  async removeByFolder(folderUuid: Folder['uuid']): Promise<number> {
+    return await this.removeByField({ folderId: folderUuid });
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:47]: Same here regarding to `async-await`

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  async findByFolder(
+    folderUuid: Folder['uuid'],
+  ): Promise<PrivateSharingFolderRole[]> {
+    const sharedFolderRolesByFolder =
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:53]: `return` directly

File: src/modules/private-share-folder/private-sharing-folder-roles.repository.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        where: { folderId: folderUuid },
+      });
+    return sharedFolderRolesByFolder;
+  }
+
+  async findByFolderAndUser(
+    folderUuid: Folder['uuid'],
+    userUuid: User['uuid'],
+  ): Promise<PrivateSharingFolderRole[]> {
+    const sharedFolderRolesByFolderAndUser =
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing-folder-roles.repository.ts:64]: `return` directly

File: src/modules/private-share-folder/private-sharing.controller.ts [hunk 18]
```diff
@@ -239,4 +245,65 @@ export class PrivateSharingController {
         ),
     };
   }
+
+  @Delete('stop/folder-id/:folderId')
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.controller.ts:249]: You are deleting a resource here, why not `DELETE /:id` ? As you are deleting a private sharing, then before deleting you match that the user id requesting this OR the `owner_id` OR the `shared_with`

File: src/modules/private-share-folder/private-sharing.controller.ts [hunk 19]
```diff
@@ -239,4 +245,65 @@ export class PrivateSharingController {
...
+        `[PRIVATESHARING/STOP] ERROR: ${
+          (error as Error).message
+        }, Error while stopping shared folder by folder ${folderUuid}, ${
+          error.stack || 'No stack trace'
+        }`,
+      );
+    }
+  }
+
+  @Delete('remove/folder-id/:folderId/user-id/:userId')
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.controller.ts:277]: The HTTP Method is DELETE, `remove` word usage then, is redundant. Also, matching with the private-sharing id as the above endpoint is enough, I think

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 20]
```diff
@@ -31,12 +41,37 @@ export class SequelizePrivateSharingRepository
...
-    @InjectModel(FolderModel)
-    private folderModel: typeof FolderModel,
     @InjectModel(PrivateSharingFolderRolesModel)
     private privateSharingFolderRole: typeof PrivateSharingFolderRolesModel,
   ) {}
 
+  private async removeByField(
+    fields: Partial<PrivateSharingFolder>,
+  ): Promise<number> {
+    const amountRemoves = await this.privateSharingFolderModel.destroy({
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.repository.ts:51]: return directly so you can remove `async-await`

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 21]
```diff
@@ -31,12 +41,37 @@ export class SequelizePrivateSharingRepository
...
+      where: fields,
+    });
+    return amountRemoves;
+  }
+
+  async removeByFolderUuid(folderUuid: string): Promise<number> {
+    const privatesRemovedByFolderUuid = await this.removeByField({
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.repository.ts:58]: return directly so you can remove `async-await`

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 22]
```diff
@@ -31,12 +41,37 @@ export class SequelizePrivateSharingRepository
...
+      folderId: folderUuid,
+    });
+    return privatesRemovedByFolderUuid;
+  }
+
+  async removeBySharedWith(
+    folderUuid: string,
+    userUuid: string,
+  ): Promise<number> {
+    const sharedWithRemoved = await this.removeByField({
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.repository.ts:68]: return directly so you can remove `async-await`

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 23]
```diff
@@ -128,4 +163,28 @@ export class SequelizePrivateSharingRepository
 
     return sharedFolders.map((folder) => folder.get({ plain: true }));
   }
+
+  async findByFolder(
+    folderUuid: Folder['uuid'],
+  ): Promise<PrivateSharingFolder[]> {
+    const sharedFolders = await this.privateSharingFolderModel.findAll({
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.repository.ts:170]: return directly so you can remove `async-await`

File: src/modules/private-share-folder/private-sharing.repository.ts [hunk 24]
```diff
@@ -128,4 +163,28 @@ export class SequelizePrivateSharingRepository
...
+      },
+    });
+    return sharedFolders;
+  }
+
+  async findByFolderAndSharedWith(
+    folderUuid: Folder['uuid'],
+    userUuid: User['uuid'],
+  ): Promise<PrivateSharingFolder[]> {
+    const sharedFolders = await this.privateSharingFolderModel.findAll({
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.repository.ts:182]: return directly so you can remove `async-await`

File: src/modules/private-share-folder/private-sharing.controller.ts [hunk 25]
```diff
@@ -239,4 +245,65 @@ export class PrivateSharingController {
...
+  @ApiOperation({
+    summary: 'Stop sharing one folder',
+  })
+  @ApiBearerAuth()
+  async stopSharing(
+    @Param('folderId', ParseUUIDPipe) folderUuid: string,
+    @UserDecorator() user: User,
+  ): Promise<any> {
+    try {
+      return await this.privateSharingUseCase.stopSharing(folderUuid);
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.controller.ts:259]: `await` here is redundant

File: src/modules/private-share-folder/private-sharing.usecase.ts [hunk 26]
```diff
@@ -70,4 +80,68 @@ export class PrivateSharingUseCase {
     );
     return folders;
   }
+
+  async stopSharing(folderUuid: Folder['uuid']): Promise<any> {
```
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.usecase.ts:84]: Well, this is called from the controller so, as far as I have a JWT authorized and I know the folder `uuid` which I want to remove **I would be able to do so, even if it is not mine, even if I am not even related to this.** 

_Security is not an option, is a must._

File: src/modules/private-share-folder/private-sharing.usecase.ts
Comment (sg-gs) [inline src/modules/private-share-folder/private-sharing.usecase.ts:1]: Check on this file that only authorized actors are able to perform the actions, please.

**Review by sg-gs — CHANGES_REQUESTED** [review]

**Review by sg-gs — APPROVED** [review]

**Pull request merged**